src/work/BasicWork.h

```diff
@@ -0,0 +1,114 @@
+// Copyright 2018 Stellar Development Foundation and contributors. Licensed
+// under the Apache License, Version 2.0. See the COPYING file at the root
+// of this distribution or at http://www.apache.org/licenses/LICENSE-2.0
+#pragma once
+
+#include "main/Application.h"
+
+namespace stellar
+{
+
+/** BasicWork is an implementation of a finite state machine,
+ * that is used for async or long-running tasks that:
+ *  - May need to be broken into steps, so as not to block the main thread
+ *  - May fail and need to retry, after some delay
+ *
+ *  BasicWork manages all state transitions via `crankWork` as well as retries.
+ *  While customers can trigger cranking and check on BasicWork's status,
+ *  it is implementers' responsibility to implement `onRun`, that
+ *  hints the next desired state.
+ */
+class BasicWork : public std::enable_shared_from_this<BasicWork>,
+                  private NonMovableOrCopyable
+{
+  public:
+    static size_t const RETRY_NEVER;
+    static size_t const RETRY_ONCE;
+    static size_t const RETRY_A_FEW;
+    static size_t const RETRY_A_LOT;
+    static size_t const RETRY_FOREVER;
+
+    enum State
```

Comment:
use c++11 syntax for enum to make it "type safe".

need a comment that explains what those states are, probably split into:
* as a user of a BasicWork, what should I expect from those states (when I query the object with `getState`)
* as an implementor, when should I return any particular value?

After writing this, it then might be more clear that we need 2 enums and not one:
* possible state a BasicWork can be, there might be only one `WORK_FAILURE`
* transitional states (used internally, such as `WORK_FAILURE_RETRY`)

Reply:
Also, wrt "terminal" states versus temporary ones, we should be consistent naming them (all the temporary states end with `-ING`, terminal states should be `DONE_XYZ`).

Reply:
Added a TODO for this, will be added in the actual implementation (since it requires changing every Work)